pkg/agent/plugin/nodeattestor/tpmdevid/tpmutil/signingkey.go

```diff
@@ -0,0 +1,180 @@
+package tpmutil
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/google/go-tpm-tools/tpm2tools"
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+	"github.com/hashicorp/go-hclog"
+	"golang.org/x/crypto/cryptobyte"
+	"golang.org/x/crypto/cryptobyte/asn1"
+)
+
+// maxAttempts indicates the max number retries for running TPM commands when
+// TPM responds with a tpm2.RCRetry code.
+const maxAttempts = 10
+
+// SigningKey represents a TPM loaded key
+type SigningKey struct {
+	Handle     tpmutil.Handle
+	sigHashAlg tpm2.Algorithm
+	rw         io.ReadWriter
+	log        hclog.Logger
+}
+
+// LoadSigningKey loads the given keypair into the provided TPM
+func LoadSigningKey(rw io.ReadWriter, pubKey, privKey []byte, log hclog.Logger) (*SigningKey, error) {
+	pub, err := tpm2.DecodePublic(pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("tpm2.Public decoding failed: %w", err)
+	}
+
+	canSign := pub.Attributes&tpm2.FlagSign != 0
+	if !canSign {
+		return nil, fmt.Errorf("not a signing key")
+	}
+
+	var sigHashAlg tpm2.Algorithm
+	var srkTemplate tpm2.Public
+	switch pub.Type {
```

Comment:
RSA and ECC have different properties, both in terms of security and performance. If someone has made an informed key choice given these properties, then I think that we probably want to use the same key type here in order to preserve that choice? Unless there is a drawback to using ECC for the AK

Relatedly, if we mix key types, then only one or the other needs to be broken in order to subvert the process, making things generally weaker than before. It could be that breaking this key doesn't subvert the entire process, but it's hard to tell from here and I can only assume that it does.

Reply:
Oh I see, thanks for the explanation, that makes a lot of sense. 

I tried to update the code to generate the attestation key of the same type of the DevID key. I found that there is some missing functionality in the gotpm library that does not allow us to do this yet.

The fix is included in the master branch of the library but it hasn't been released yet. I updated the docs to reflect this situation (attestation key is always RSA).